### DISCUSSION TITLE
Is it possible to switch pnpm version back and forth like nvm?

### DISCUSSION THREAD

**Discussion by jonathanfilbert** [pnpm/pnpm#4446] [Q&A] [answered]
Hey, so my company is migrating from `pnpm@4` to `pnpm@6` and since we have a lot of services, some already migrated into `pnpm@6` whilst some are still in `pnpm@4`. Let's say for example my machine is running on `pnpm@4` and I have to develop a feature on a service that's running `pnpm@6`, I would always get an error like the following:
```
Expected version: >=6.23.2
Got: 4.14.4
```
the same applies if I'm running `pnpm@6` and want to develop a feature on a service running `pnpm@4`.

Is there some kind of an [nvm](https://github.com/nvm-sh/nvm#intro) where we can switch quickly between pnpm version to a specific version needed for our codebase? If not, what is the recommended strategy for teams to develop features in-between pnpm transition?

Thanks!

**Comment by zkochan** [comment, marked as answer]
Sorry, I am currently not aware of such feature. So you probably need to reinstall pnpm. Or you can dynamically run pnpm.

To run v4:

```
pnpm dlx pnpm@4 install
```

To run v6:

```
pnpm dlx pnpm@6 install
```

Also, v7 is already in progress. It will be latest on May 1st

**Comment by Bessonov** [comment]
May be you are interested in [this answer](https://github.com/pnpm/pnpm/issues/1383#issuecomment-1152273326) about corepack.

**Comment by airtonix** [comment]
use asdf instead of nvm, then you can manage all kinds of tools (we use it for pnpm, node, yarn etc)

**Comment by zkochan** [comment]
It is now possible with [pnpm v9.7](https://github.com/pnpm/pnpm/releases/tag/v9.7.0).

**Reply by ggascoigne** [reply]
Honestly I don't want corepack (it'd be better if it was always installed and always enabled, but it isn't), I also don't really want pnpm to auto-update.  I work on a number of projects and I want to lock down build tool versions for each project, if one team is on pnpm 9.7 and another is on 10.18 then I want to be able to install both, and then switch between then and run them.

asdf does this and so far is the best solution I've found.

**Reply by zkochan** [reply]
> I work on a number of projects and I want to lock down build tool versions for each project, if one team is on pnpm 9.7 and another is on 10.18 then I want to be able to install both, and then switch between then and run them.

This works with pnpm out of the box. No need in corepack or asdf.

**Reply by ggascoigne** [reply]
Specifically how? something like `pnpm dlx pmpn@9.7.1 run command`, or something else?

**Reply by zkochan** [reply]
pnpm will read the "packageManager" field in "package.json" and automatically switch to the version specified in it.

If you want to switch to something else, you can run `pnpm self-update <version>`.

**Reply by ggascoigne** [reply]
nice, that's a nice way to handle it.  I always found corepack a pain because it wasn't generally enabled.  This is much better.
